cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp

```diff
@@ -0,0 +1,389 @@
+    {
+        out_block.info.is_overflows = aggregate_info->is_overflows;
+        out_block.info.bucket_num = aggregate_info->bucket_num;
+    }
+    partition_writer->write(partition_info, out_block);
+    split_result.wall_time += wall_time.elapsedNanoseconds();
+}
+
+void SparkExchangeSink::onFinish()
+{
+    Stopwatch wall_time;
+    if (!dynamic_cast<LocalPartitionWriter*>(partition_writer.get()))
+    {
+        partition_writer->evictPartitions();
+    }
+    split_result.wall_time += wall_time.elapsedNanoseconds();
+}
+
+void SparkExchangeSink::initOutputHeader(const Block & block)
+{
+    if (!output_header)
+    {
+        if (output_columns_indicies.empty())
+        {
+            output_header = block.cloneEmpty();
+            for (size_t i = 0; i < block.columns(); ++i)
+                output_columns_indicies.push_back(i);
+        }
+        else
+        {
+            ColumnsWithTypeAndName cols;
+            for (const auto & index : output_columns_indicies)
+                cols.push_back(block.getByPosition(index));
+
+            output_header = Block(std::move(cols));
+        }
+    }
+}
+
+SparkExchangeManager::SparkExchangeManager(const Block& header, const String & short_name, const SplitOptions & options_, jobject rss_pusher): input_header(materializeBlock(header)), options(options_)
+{
+    if (rss_pusher)
+    {
+        GET_JNIENV(env)
+        jclass celeborn_partition_pusher_class =
+            CreateGlobalClassReference(env, "Lorg/apache/spark/shuffle/CelebornPartitionPusher;");
+        jmethodID celeborn_push_partition_data_method =
+            GetMethodID(env, celeborn_partition_pusher_class, "pushPartitionData", "(I[BI)I");
+        CLEAN_JNIENV
+        celeborn_client = std::make_unique<CelebornClient>(rss_pusher, celeborn_push_partition_data_method);
+        use_rss = true;
+    }
+    if (!partitioner_creators.contains(short_name))
+    {
+        throw DB::Exception(DB::ErrorCodes::BAD_ARGUMENTS, "unsupported splitter {}", short_name);
+    }
+    partitioner_creator = partitioner_creators[short_name];
+    Poco::StringTokenizer output_column_tokenizer(options_.out_exprs, ",");
+    for (const auto & iter : output_column_tokenizer)
+        output_columns_indicies.push_back(std::stoi(iter));
+    auto overhead_memory = header.columns() * 16 * options.split_size * options.partition_num;
+    use_sort_shuffle = overhead_memory > options.spill_threshold * 0.5 || options.partition_num >= 300 || options.force_memory_sort;
+
+    split_result.partition_lengths.resize(options.partition_num, 0);
+    split_result.raw_partition_lengths.resize(options.partition_num, 0);
+}
+
+std::shared_ptr<PartitionWriter> createPartitionWriter(const SplitOptions& options, bool use_sort_shuffle, std::unique_ptr<CelebornClient> celeborn_client)
+{
+    if (celeborn_client)
+    {
+        if (use_sort_shuffle)
+            return std::make_shared<MemorySortCelebornPartitionWriter>(options, std::move(celeborn_client));
```

Comment:
I wonder whether a celeborn client could be shared among multi threads. If it's sharable, make it a shared_ptr may be better.

Reply:
celeborn will not have any performance benefits in parallel, and will later resize the pipeline to ensure that there is only one sink